hnn_core/extracellular.py

```diff
@@ -0,0 +1,601 @@
+
+        The handler is set up to maintain a vector of membrane currents at at
+        every inner segment of every section of every cell on each CVODE
+        integration step. In addition, it records a time vector of sample
+        times.
+
+        Parameters
+        ----------
+        cvode : instance of h.CVode
+            Multi order variable time step integration method.
+        """
+        # ordered list of h.Sections on this rank (if running in parallel)
+        secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        segment_counts = np.array(_get_segment_counts(secs_on_rank))
+        n_total_segments = segment_counts.sum()
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self._nrn_imem_ptrvec = h.PtrVector(n_total_segments)
+        # placeholder into which pointer values are read on each sim time step
+        self._nrn_imem_vec = h.Vector(n_total_segments)
+
+        ptr_count = 0
+        for sec in secs_on_rank:
+            for seg in sec:  # section end points (0, 1) not included
+                # set Nth pointer to the net membrane current at this segment
+                self._nrn_imem_ptrvec.pset(
+                    ptr_count, sec(seg.x)._ref_i_membrane_)
+                ptr_count += 1
+        if ptr_count != n_total_segments:
+            raise RuntimeError(f'Expected {n_total_segments} imem pointers, '
+                               f'got {ptr_count}.')
+
+        # transfer resistances for each segment (keep in Neuron Matrix object)
+        self._nrn_r_transfer = h.Matrix(self.n_contacts, n_total_segments)
+
+        for row, pos in enumerate(self.array.positions):
+            if self.array.method is not None:
+                transfer_resistance = list()
+                for sec in secs_on_rank:
+                    this_xfer_r = _transfer_resistance(
+                        sec, pos, sigma=self.array.sigma,
+                        method=self.array.method,
+                        min_distance=self.array.min_distance)
+                    transfer_resistance.extend(this_xfer_r)
+
+                self._nrn_r_transfer.setrow(row, h.Vector(transfer_resistance))
+            else:
+                # for testing, make a matrix of ones
+                self._nrn_r_transfer.setrow(row,
+                                            h.Vector(segment_counts.sum(), 1.))
+
+        # record time for each array
+        self._nrn_times = h.Vector().record(h._ref_t)
+
+        # contributions of all segments on this rank to total calculated
+        # potential at electrode (_PC.allreduce called in _simulate_dipole)
+        self._nrn_voltages = h.Vector()
+        self._nrn_imem_vec = h.Vector(n_total_segments)
+
+        # Attach a callback for calculating the potentials at each time step.
+        # Enables fast calculation of transmembrane current (nA) at each
```

Comment:
is the "fast" part of this really true?

Reply:
I don't have any numbers for you, but will stand by a firm: yes :) Note that _everything_ in the callback remains in Neuron/C, so we don't waste cycles on creating Python objects on each iteration!